src/convert/mod.rs

```diff
@@ -0,0 +1,264 @@
+}
+
+impl SequenceLength {
+    pub fn new<S: ToString>(name: &S, length: usize) -> Self {
+        Self {
+            name: name.to_string(),
+            length,
+        }
+    }
+
+    #[must_use]
+    pub const fn len(&self) -> usize {
+        self.length
+    }
+
+    #[must_use]
+    pub const fn is_empty(&self) -> bool {
+        self.length == 0
+    }
+}
+
+/// Trait for readers of other file formats
+///
+/// This trait is used in the function [`to_2bit`](fn.to_2bit.html) to
+/// retrieve the needed information to produce a 2bit file from another file format.
+///
+/// The trait has several methods that return a `Cow`. This `Cow` wrapper is intended to give
+/// implementors the flexibility of either returning an owned or borrowed result.
+/// Depending on the underlying file format, calculating results on the fly might be more effective
+/// than returning a reference to a cached result.
+pub trait SequenceRead<'a> {
+    /// Return the names and lengths of the sequences (ideally in a stable order for reproducibility)
+    fn sequence_lengths(&'a self) -> Cow<'a, [SequenceLength]>;
+    /// Return pieces of the actual nucleotide sequence
+    fn nucleotides(&self, chr: &str) -> Box<dyn Nucleotides>;
+    /// Return the soft masked blocks for a sequence
+    fn soft_masked_blocks(&'a self, chr: &str) -> Cow<'a, [Block]>;
+    /// Return the hard masked blocks for a sequence
+    fn hard_masked_blocks(&'a self, chr: &str) -> Cow<'a, [Block]>;
+}
+
+/// Trait for providers of nucleotide sequence data.
+///
+/// The data is typically provided in chunks to avoid allocating too long sequences.
+pub trait Nucleotides {
+    /// Read the next chunk of nucleotides from the sequence.
+    /// The underlying reader will place the nucleotides into the provided
+    /// buffer until the buffer is full.
+    /// If the sequence is longer than the buffer, then the consumer needs to
+    /// call the method again to retrieve more nucleotides.
+    ///
+    /// This method returns the number of Nucleotides placed into the buffer or
+    /// `None` if the underlying reader has reached the end of the sequence.
+    fn read_chunk(
+        &mut self,
+        buf: &mut [Nucleotide],
+    ) -> std::result::Result<Option<usize>, Box<dyn std::error::Error>>;
+}
+
+/// Convert and write data from a `SequenceRead` into a new 2bit file.
+#[allow(clippy::too_many_lines)]
+pub fn to_2bit<'a>(
+    writer: &mut dyn Write,
+    extractor: &'a (dyn SequenceRead<'a> + 'a),
+) -> Result<()> {
+    let seqs = extractor.sequence_lengths();
+
+    let signature: Field = SIGNATURE;
+    let version: Field = 0;
+    let sequence_count: Field = seqs.len() as Field;
+    let reserved: Field = 0;
+
+    // let's start writing immediately to force IO errors early
+    writer.write_all(&signature.to_ne_bytes())?;
+    writer.write_all(&version.to_ne_bytes())?;
+    writer.write_all(&sequence_count.to_ne_bytes())?;
+    writer.write_all(&reserved.to_ne_bytes())?;
```

Comment:
```suggestion
    for input in &[&signature, &version, sequence, sequence_count, reserved] {
        writer_write_all(input.to_ne_bytes())?;
    }
```

Reply:
That's a good idea! Thanks!

Reply:
Oops. The `sequence` variable should not have been part of that list. I'll fix that later.